### PR TITLE
Allow to set custom WS URL waiting time

### PR DESCRIPTION
**Pull request by karelbilek**
With a heavy CPU utilization, Chrome sometimes does not start in 20 seconds.

This adds a custom option to set the WS URL waiting time.

### CODE REVIEW

**Comment by ZekeLu** [comment]
Suggested changes:

1. Add a commit to change the arrangement of the `ExecAllocator` struct. `wg` works like a [mutext hat](https://dmitri.shuralyov.com/idiomatic-go/entries/2) here.

```diff
diff --git a/allocate.go b/allocate.go
index f639ea2..3de99b1 100644
--- a/allocate.go
+++ b/allocate.go
@@ -105,11 +105,11 @@ type ExecAllocator struct {
    initFlags map[string]interface{}
    initEnv   []string
 
+   modifyCmdFunc func(cmd *exec.Cmd)
+
    wg sync.WaitGroup
 
    combinedOutputWriter io.Writer
-
-   modifyCmdFunc func(cmd *exec.Cmd)
 }
```

2. Add another commit to implement `WSURLReadTimeout`:

```diff
diff --git a/allocate.go b/allocate.go
index 3de99b1..1958308 100644
--- a/allocate.go
+++ b/allocate.go
@@ -37,7 +37,8 @@ type Allocator interface {
 // to create the allocator without the unnecessary context layer.
 func setupExecAllocator(opts ...ExecAllocatorOption) *ExecAllocator {
    ep := &ExecAllocator{
-       initFlags: make(map[string]interface{}),
+       initFlags:        make(map[string]interface{}),
+       wsURLReadTimeout: 20 * time.Second,
    }
    for _, o := range opts {
        o(ep)
@@ -107,6 +108,11 @@ type ExecAllocator struct {
 
    modifyCmdFunc func(cmd *exec.Cmd)
 
+   // Chrome will sometimes fail to print the websocket, or run for a long
+   // time, without properly exiting. To avoid blocking forever in those
+   // cases, give up after a specified timeout.
+   wsURLReadTimeout time.Duration
+
    wg sync.WaitGroup
 
    combinedOutputWriter io.Writer
@@ -224,11 +230,6 @@ func (a *ExecAllocator) Allocate(ctx context.Context, opts ...BrowserOption) (*B
        close(c.allocated)
    }()
 
-   // Chrome will sometimes fail to print the websocket, or run for a long
-   // time, without properly exiting. To avoid blocking forever in those
-   // cases, give up after twenty seconds.
-   const wsURLReadTimeout = 20 * time.Second
-
    var wsURL string
    wsURLChan := make(chan struct{}, 1)
    go func() {
@@ -237,7 +238,7 @@ func (a *ExecAllocator) Allocate(ctx context.Context, opts ...BrowserOption) (*B
    }()
    select {
    case <-wsURLChan:
-   case <-time.After(wsURLReadTimeout):
+   case <-time.After(a.wsURLReadTimeout):
        err = errors.New("websocket url timeout reached")
    }
    if err != nil {
@@ -494,6 +495,14 @@ func CombinedOutput(w io.Writer) ExecAllocatorOption {
    }
 }
 
+// WSURLReadTimeout sets the waiting time for reading the WebSocket URL.
+// The default value is 20 seconds.
+func WSURLReadTimeout(t time.Duration) ExecAllocatorOption {
+   return func(a *ExecAllocator) {
+       a.wsURLReadTimeout = t
+   }
+}
+
 // NewRemoteAllocator creates a new context set up with a RemoteAllocator,
 // suitable for use with NewContext. The url should point to the browser's
 // websocket address, such as "ws://127.0.0.1:$PORT/devtools/browser/...".
```

**Comment by karelbilek** [comment]
Will do

**Comment by karelbilek** [comment]
Commit updated..

File: allocate.go [hunk 1]
```diff
@@ -105,11 +106,15 @@ type ExecAllocator struct {
 	initFlags map[string]interface{}
 	initEnv   []string
 
-	wg sync.WaitGroup
+	modifyCmdFunc func(cmd *exec.Cmd)
```
Comment (ZekeLu) [inline allocate.go:109]: Please make this change into its own commit. And explain why the change.

File: allocate.go [hunk 2]
```diff
@@ -105,11 +106,15 @@ type ExecAllocator struct {
...
 
-	combinedOutputWriter io.Writer
+	// Chrome will sometimes fail to print the websocket, or run for a long
+	// time, without properly exiting. To avoid blocking forever in those
+	// cases, give up after a specified timeout.
+	wsURLReadTimeout time.Duration
 
-	modifyCmdFunc func(cmd *exec.Cmd)
+	wg                   sync.WaitGroup
```
Comment (ZekeLu) [inline allocate.go:116]: Do not remove the blank line after `wg sync.WaitGroup` because `wg` is not totally for `combinedOutputWriter`.

**Review by ZekeLu — CHANGES_REQUESTED** [review]

File: allocate.go (hunk 1 above)
Comment (ZekeLu) [reply allocate.go:109]: See this comment: https://github.com/chromedp/chromedp/pull/1036#issuecomment-1067839839

File: allocate.go (hunk 2 above)
Comment (ZekeLu) [reply allocate.go:116]: reference:

https://github.com/chromedp/chromedp/blob/621263b43cbf8797c256ffac2508d955f3f8ad95/allocate.go#L206-L209

**Comment by waybackarchiver** [comment]
Any news?

File: allocate.go (hunk 2 above)
Comment (karelbilek) [reply allocate.go:116]: II guess I  just do not understand your request 1. sorry :( 

You said wg works like a "mutex hat". Reading about mutex hats, from your link - "It sits, like a hat, on top of the variables that it protects." 

What does it protect here?

I just don't understand what is the purpose of moving `modifyCmdFunc func(cmd *exec.Cmd)` up. And if it should be a separate commit, what should be reason for that in that commit.

**Comment by karelbilek** [comment]
@waybackarchiver hey. I realized this is a bit nonsensical; when the CPU usage is high, and chrome does not start in 20 seconds, in my experience, it still doesn't start in 1 minute or 2 minutes or ever; so this PR is a bit useless.

I still think it should be user-settable though, as the 20 seconds timeout is a bit random and should not be hardcoded; but I don't have it as priority 1 to merge this :)

File: allocate.go (hunk 2 above)
Comment (ZekeLu) [reply allocate.go:116]: Sorry, it's my fault that I haven't made it clear. First of all, this is a personal taste. So it's a little hard to understand. I will try to explain it.

First, see the source code below:

https://github.com/chromedp/chromedp/blob/621263b43cbf8797c256ffac2508d955f3f8ad95/allocate.go#L206-L225

https://github.com/chromedp/chromedp/blob/621263b43cbf8797c256ffac2508d955f3f8ad95/allocate.go#L234-L237

https://github.com/chromedp/chromedp/blob/621263b43cbf8797c256ffac2508d955f3f8ad95/allocate.go#L244-L248

As the comments in line 206 and 208 said, `wg` is for both the entire allocator and the io.Copy in a separate goroutine. That's why I said `wg` **works like a mutex hat**. Strictly speaking, It does not protect `combinedOutputWriter`, but it has something to do with `combinedOutputWriter`. So I think they should be arranged together in the struct.

On the other hand, `wg` is not totally for `combinedOutputWriter` (it's for the entire allocator too). If we remove the blank line between them, it will make it looks like that `wg` is totally for `combinedOutputWriter`.

So I think the best choice is to put `wg` and `combinedOutputWriter` into the last part of the struct and leave a blank line between them. This is what it's arranged before `modifyCmdFunc func(cmd *exec.Cmd)` is added in #674.

This suggested change has nothing to do with `wsURLReadTimeout`, and it's hard to explain. That's why it deserves a standalone commit.

**Comment by 8tomat8** [comment]
@karelbilek are you planning to finish this PR or should we take it over?

File: allocate.go (hunk 2 above)
Comment (8tomat8) [reply allocate.go:116]: @ZekeLu If this change has nothing to do with `wsURLReadTimeout`, then why it is even discussed in this PR?

**Comment by ubombi** [comment]
I lambda, during cold start is occasionally takes like ~30 sec or more. And start each time, having this hard-coded artificial timeout removed from the source code.

File: allocate.go (hunk 2 above)
Comment (ubombi) [reply allocate.go:116]: @8tomat8 
```
+// WSURLReadTimeout sets the waiting time for reading the WebSocket URL.
+// The default value is 20 seconds.
+func WSURLReadTimeout(t time.Duration) ExecAllocatorOption {
+   return func(a *ExecAllocator) {
+       a.wsURLReadTimeout = t
+   }
+}
+
```

Comment (ZekeLu) [reply allocate.go:116]: > @ZekeLu If this change has nothing to do with `wsURLReadTimeout`, then why it is even discussed in this PR?

I have submitted 1c2b99f so we can ignore this part now.

**Comment by ZekeLu** [comment]
I'm going to cherry-pick the changes.

**Comment by ZekeLu** [comment]
done.

**Comment by 8tomat8** [comment]
@ZekeLu When is the next release planned? I want to plan switch on this repo from our fork

**Comment by ZekeLu** [comment]
@8tomat8 `v0.8.4` has just been released.
